com.hazelcast.jet.job.JobSubmissionSlownessRegressionTest.regressionTestForPR1488

_master_ / _3.2-maintenance_ (commit 166503c66113d5fe3d7e3158f4aba87cc683a7b9)

Failed on nightly build: http://jenkins.hazelcast.com/job/jet-oss-maintenance-nightly/52/testReport/junit/com.hazelcast.jet.job/JobSubmissionSlownessRegressionTest/regressionTestForPR1488/

**It fails intermittently and it is caused by nature of the test since it is not deterministic. Fix will be probably very tricky and will probably need redesign of whole test.**

Stacktrace:
```
java.lang.AssertionError: Job submission rate should not decrease. First rate: 5.197103105280812, second rate: 3.840073037683318
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.jet.job.JobSubmissionSlownessRegressionTest.regressionTestForPR1488(JobSubmissionSlownessRegressionTest.java:120)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

Standard output:
```
Hiccups measured while running test 'regressionTestForPR1488(com.hazelcast.jet.job.JobSubmissionSlownessRegressionTest):'
19:16:50, accumulated pauses: 984 ms, max pause: 616 ms, pauses over 1000 ms: 0
19:16:55, accumulated pauses: 391 ms, max pause: 101 ms, pauses over 1000 ms: 0
19:17:00, accumulated pauses: 46 ms, max pause: 3 ms, pauses over 1000 ms: 0
19:17:05, accumulated pauses: 1449 ms, max pause: 1122 ms, pauses over 1000 ms: 1
19:17:10, accumulated pauses: 120 ms, max pause: 18 ms, pauses over 1000 ms: 0
19:17:15, accumulated pauses: 83 ms, max pause: 33 ms, pauses over 1000 ms: 0
19:17:20, accumulated pauses: 278 ms, max pause: 87 ms, pauses over 1000 ms: 0
19:17:25, accumulated pauses: 88 ms, max pause: 55 ms, pauses over 1000 ms: 0
19:17:30, accumulated pauses: 367 ms, max pause: 190 ms, pauses over 1000 ms: 0
19:17:35, accumulated pauses: 225 ms, max pause: 116 ms, pauses over 1000 ms: 0
19:17:40, accumulated pauses: 93 ms, max pause: 45 ms, pauses over 1000 ms: 0
19:17:45, accumulated pauses: 36 ms, max pause: 1 ms, pauses over 1000 ms: 0
19:17:50, accumulated pauses: 399 ms, max pause: 121 ms, pauses over 1000 ms: 0
19:17:55, accumulated pauses: 132 ms, max pause: 19 ms, pauses over 1000 ms: 0
19:18:00, accumulated pauses: 424 ms, max pause: 314 ms, pauses over 1000 ms: 0
19:18:05, accumulated pauses: 272 ms, max pause: 67 ms, pauses over 1000 ms: 0
19:18:10, accumulated pauses: 56 ms, max pause: 13 ms, pauses over 1000 ms: 0
19:18:15, accumulated pauses: 236 ms, max pause: 64 ms, pauses over 1000 ms: 0
19:18:20, accumulated pauses: 267 ms, max pause: 93 ms, pauses over 1000 ms: 0
19:18:25, accumulated pauses: 62 ms, max pause: 31 ms, pauses over 1000 ms: 0
19:18:30, accumulated pauses: 709 ms, max pause: 626 ms, pauses over 1000 ms: 0
19:18:35, accumulated pauses: 232 ms, max pause: 45 ms, pauses over 1000 ms: 0
19:18:40, accumulated pauses: 76 ms, max pause: 39 ms, pauses over 1000 ms: 0
19:18:45, accumulated pauses: 184 ms, max pause: 67 ms, pauses over 1000 ms: 0
19:18:50, accumulated pauses: 194 ms, max pause: 68 ms, pauses over 1000 ms: 0
19:18:55, accumulated pauses: 682 ms, max pause: 469 ms, pauses over 1000 ms: 0
19:19:00, accumulated pauses: 103 ms, max pause: 31 ms, pauses over 1000 ms: 0
19:19:05, accumulated pauses: 249 ms, max pause: 56 ms, pauses over 1000 ms: 0
19:19:10, accumulated pauses: 185 ms, max pause: 57 ms, pauses over 1000 ms: 0
19:19:15, accumulated pauses: 85 ms, max pause: 48 ms, pauses over 1000 ms: 0
19:19:20, accumulated pauses: 295 ms, max pause: 73 ms, pauses over 1000 ms: 0
19:19:25, accumulated pauses: 935 ms, max pause: 828 ms, pauses over 1000 ms: 0
19:19:30, accumulated pauses: 244 ms, max pause: 43 ms, pauses over 1000 ms: 0
19:19:35, accumulated pauses: 36 ms, max pause: 2 ms, pauses over 1000 ms: 0
19:19:40, accumulated pauses: 330 ms, max pause: 99 ms, pauses over 1000 ms: 0
19:19:45, accumulated pauses: 104 ms, max pause: 17 ms, pauses over 1000 ms: 0
19:19:50, accumulated pauses: 36 ms, max pause: 0 ms, pauses over 1000 ms: 0
19:19:55, accumulated pauses: 710 ms, max pause: 570 ms, pauses over 1000 ms: 0
19:20:00, accumulated pauses: 381 ms, max pause: 106 ms, pauses over 1000 ms: 0
19:20:05, accumulated pauses: 66 ms, max pause: 12 ms, pauses over 1000 ms: 0
19:20:10, accumulated pauses: 60 ms, max pause: 13 ms, pauses over 1000 ms: 0
19:20:15, accumulated pauses: 688 ms, max pause: 561 ms, pauses over 1000 ms: 0
19:20:20, accumulated pauses: 44 ms, max pause: 2 ms, pauses over 1000 ms: 0
19:20:25, accumulated pauses: 51 ms, max pause: 3 ms, pauses over 1000 ms: 0
19:20:30, accumulated pauses: 40 ms, max pause: 3 ms, pauses over 1000 ms: 0
19:20:35, accumulated pauses: 341 ms, max pause: 122 ms, pauses over 1000 ms: 0
19:20:40, accumulated pauses: 441 ms, max pause: 190 ms, pauses over 1000 ms: 0
19:20:45, accumulated pauses: 37 ms, max pause: 0 ms, pauses over 1000 ms: 0
19:20:50, accumulated pauses: 51 ms, max pause: 11 ms, pauses over 1000 ms: 0
19:20:55, accumulated pauses: 349 ms, max pause: 202 ms, pauses over 1000 ms: 0
19:21:00, accumulated pauses: 270 ms, max pause: 134 ms, pauses over 1000 ms: 0
19:21:05, accumulated pauses: 118 ms, max pause: 60 ms, pauses over 1000 ms: 0
19:21:10, accumulated pauses: 633 ms, max pause: 261 ms, pauses over 1000 ms: 0
19:21:15, accumulated pauses: 116 ms, max pause: 19 ms, pauses over 1000 ms: 0
19:21:20, accumulated pauses: 334 ms, max pause: 260 ms, pauses over 1000 ms: 0
19:21:25, accumulated pauses: 42 ms, max pause: 1 ms, pauses over 1000 ms: 0
19:21:30, accumulated pauses: 457 ms, max pause: 219 ms, pauses over 1000 ms: 0
19:21:35, accumulated pauses: 276 ms, max pause: 43 ms, pauses over 1000 ms: 0
19:21:40, accumulated pauses: 83 ms, max pause: 39 ms, pauses over 1000 ms: 0
19:21:45, accumulated pauses: 34 ms, max pause: 1 ms, pauses over 1000 ms: 0
19:21:50, accumulated pauses: 277 ms, max pause: 117 ms, pauses over 1000 ms: 0
19:21:55, accumulated pauses: 251 ms, max pause: 58 ms, pauses over 1000 ms: 0
19:22:00, accumulated pauses: 64 ms, max pause: 5 ms, pauses over 1000 ms: 0
19:22:05, accumulated pauses: 79 ms, max pause: 22 ms, pauses over 1000 ms: 0
19:22:10, accumulated pauses: 1194 ms, max pause: 833 ms, pauses over 1000 ms: 0
19:22:15, accumulated pauses: 40 ms, max pause: 3 ms, pauses over 1000 ms: 0
19:22:20, accumulated pauses: 71 ms, max pause: 19 ms, pauses over 1000 ms: 0
19:22:25, accumulated pauses: 149 ms, max pause: 70 ms, pauses over 1000 ms: 0
19:22:30, accumulated pauses: 450 ms, max pause: 212 ms, pauses over 1000 ms: 0
19:22:35, accumulated pauses: 116 ms, max pause: 18 ms, pauses over 1000 ms: 0
19:22:40, accumulated pauses: 1004 ms, max pause: 567 ms, pauses over 1000 ms: 0
19:22:45, accumulated pauses: 197 ms, max pause: 17 ms, pauses over 1000 ms: 0
19:22:50, accumulated pauses: 208 ms, max pause: 171 ms, pauses over 1000 ms: 0
19:22:55, accumulated pauses: 385 ms, max pause: 126 ms, pauses over 1000 ms: 0
19:23:00, accumulated pauses: 242 ms, max pause: 141 ms, pauses over 1000 ms: 0
19:23:05, accumulated pauses: 95 ms, max pause: 19 ms, pauses over 1000 ms: 0
19:23:10, accumulated pauses: 821 ms, max pause: 729 ms, pauses over 1000 ms: 0
19:23:15, accumulated pauses: 243 ms, max pause: 100 ms, pauses over 1000 ms: 0
19:23:20, accumulated pauses: 171 ms, max pause: 12 ms, pauses over 1000 ms: 0
19:23:25, accumulated pauses: 177 ms, max pause: 78 ms, pauses over 1000 ms: 0
19:23:30, accumulated pauses: 157 ms, max pause: 35 ms, pauses over 1000 ms: 0
19:23:35, accumulated pauses: 312 ms, max pause: 158 ms, pauses over 1000 ms: 0
19:23:40, accumulated pauses: 287 ms, max pause: 196 ms, pauses over 1000 ms: 0
19:23:45, accumulated pauses: 176 ms, max pause: 100 ms, pauses over 1000 ms: 0
19:23:50, accumulated pauses: 299 ms, max pause: 95 ms, pauses over 1000 ms: 0
19:23:55, accumulated pauses: 598 ms, max pause: 559 ms, pauses over 1000 ms: 0
```

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/job/JobSubmissionSlownessRegressionTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/job/JobSubmissionSlownessRegressionTest.java
@@ -34,27 +34,27 @@ import org.junit.runner.RunWith;
 
 import java.util.ArrayList;
 import java.util.List;
-import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
-import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import static com.hazelcast.jet.core.processor.Processors.noopP;
+import static com.hazelcast.jet.impl.util.Util.uncheckRun;
+import static java.util.concurrent.TimeUnit.SECONDS;
 import static org.junit.Assert.assertTrue;
 
 @RunWith(HazelcastSerialClassRunner.class)
 @Category({SlowTest.class, IgnoredForCoverage.class})
 public final class JobSubmissionSlownessRegressionTest extends JetTestSupport {
 
-    private static final int DURATION_SECS = 60;
-    private static final int THREADS_COUNT = Runtime.getRuntime().availableProcessors();
+    private static final int DURATION_SECS = 10;
+    private static final int THREADS_COUNT = 5;
 
-    private static final int HEAT_UP_CYCLE_COUNT = 1;
-    private static final int MEASUREMENT_A_CYCLE_COUNT = 1;
-    private static final int WAIT_BEFORE_MEASUREMENT_B_COUNT = 2;
-    private static final int MEASUREMENT_B_CYCLE_COUNT = 1;
+    private static final int HEAT_UP_CYCLE_COUNT = 3;
+    private static final int MEASUREMENT_A_CYCLE_COUNT = 3;
+    private static final int WAIT_BEFORE_MEASUREMENT_B_COUNT = 3;
+    private static final int MEASUREMENT_B_CYCLE_COUNT = 3;
 
     private static final int HEAT_UP_CYCLE_SECTION = HEAT_UP_CYCLE_COUNT;
     private static final int MEASUREMENT_A_CYCLE_SECTION = HEAT_UP_CYCLE_SECTION + MEASUREMENT_A_CYCLE_COUNT;
@@ -72,7 +72,9 @@ public final class JobSubmissionSlownessRegressionTest extends JetTestSupport {
     }
 
     @Test
-    public void regressionTestForPR1488() throws ExecutionException, InterruptedException {
+    public void regressionTestForPR1488() {
+        logger.info(String.format("Starting test with %d threads", THREADS_COUNT));
+
         ExecutorService executorService = Executors.newFixedThreadPool(THREADS_COUNT);
 
         double measurementARateSum = 0;
@@ -80,36 +82,30 @@ public final class JobSubmissionSlownessRegressionTest extends JetTestSupport {
 
         DAG dag = twoVertex();
         JetInstance client = createJetClient();
-        while (measurementCount <= MEASUREMENT_B_CYCLE_SECTION) {
-            AtomicInteger completedRoundtrips = new AtomicInteger();
+        while (measurementCount < MEASUREMENT_B_CYCLE_SECTION) {
+            AtomicInteger completedRoundTrips = new AtomicInteger();
             long start = System.nanoTime();
-            List<Future> futures = new ArrayList<Future>();
+            List<Future<?>> futures = new ArrayList<>();
             for (int i = 0; i < THREADS_COUNT; i++) {
                 Future<?> f = executorService.submit(() -> {
-                    bench(DURATION_SECS, () -> {
+                    bench(() -> {
                         client.newJob(dag, new JobConfig()).join();
-                    }, completedRoundtrips);
+                    }, completedRoundTrips);
                 });
                 futures.add(f);
             }
-            for (Future future : futures) {
-                future.get();
-            }
+            futures.forEach(f -> uncheckRun(f::get));
+
             long elapsed = System.nanoTime() - start;
-            double rate = (double) completedRoundtrips.get() / (double) elapsed * TimeUnit.SECONDS.toNanos(1);
+            double rate = (double) completedRoundTrips.get() / (double) elapsed * SECONDS.toNanos(1);
             System.out.println("Rate was " + rate + " req/s");
             measurementCount++;
-            if (measurementCount <= HEAT_UP_CYCLE_SECTION) {
-            } else {
-                if (measurementCount <= MEASUREMENT_A_CYCLE_SECTION) {
+
+            if (measurementCount > HEAT_UP_CYCLE_SECTION) { //3
+                if (measurementCount <= MEASUREMENT_A_CYCLE_SECTION) { //6
                     measurementARateSum += rate;
-                } else {
-                    if (measurementCount <= WAIT_BEFORE_MEASUREMENT_B_SECTION) {
-                    } else {
-                        if (measurementCount <= MEASUREMENT_B_CYCLE_SECTION) {
-                            measurementBRateSum += rate;
-                        }
-                    }
+                } else if (measurementCount > WAIT_BEFORE_MEASUREMENT_B_SECTION) { //9
+                    measurementBRateSum += rate;
                 }
             }
         }
@@ -129,9 +125,9 @@ public final class JobSubmissionSlownessRegressionTest extends JetTestSupport {
         return dag;
     }
 
-    private static void bench(long durationSecs, Runnable r, AtomicInteger completedRoundTrips) {
+    private static void bench(Runnable r, AtomicInteger completedRoundTrips) {
         long start = System.nanoTime();
-        long end = start + TimeUnit.SECONDS.toNanos(durationSecs);
+        long end = start + SECONDS.toNanos(DURATION_SECS);
         while (System.nanoTime() < end) {
             r.run();
             completedRoundTrips.incrementAndGet();

